lib/upgrade/etcd/conversionv1v3/names.go

```diff
@@ -15,55 +15,44 @@
 package conversionv1v3
 
 import (
+	"crypto/sha1"
+	"encoding/hex"
 	"fmt"
 	"net"
-	"regexp"
 	"strings"
-)
 
-var (
-	nonNameChar               = regexp.MustCompile("[^-.a-z0-9]+")
-	nonNameNoDotChar          = regexp.MustCompile("[^-a-z0-9]+")
-	dotDashSeq                = regexp.MustCompile("[.-]*[.][.-]*")
-	trailingLeadingDotsDashes = regexp.MustCompile("^[.-]*(.*?)[.-]*$")
+	"github.com/projectcalico/libcalico-go/lib/names"
 )
 
-// Convert the v1 name to a standard v3 name.  This converts as follows (in the
-// listed order):
-// -  Convert to lowercase
-// -  Convert [/] to .
-// -  Convert any other char that is not in the set [-.a-z0-9] to -
-// -  Convert any multi-byte sequence of [-.] with at least one [.] to a single .
-// -  Remove leading and trailing dashes and dots
+// Convert the v1 name to a standard v3 name.  This uses the standard name normalization,
+// and adds an additional qualifier if the name was modified.  The qualifier is calculated
+// from the original name.
 func convertName(v1Name string) string {
-	name := strings.ToLower(v1Name)
-	name = strings.Replace(name, "/", ".", -1)
-	name = nonNameChar.ReplaceAllString(name, "-")
-	name = dotDashSeq.ReplaceAllString(name, ".")
+	name := names.NormalizeName(v1Name)
 
-	// Extract the trailing and leading dots and dashes.   This should always match even if
-	// the matched substring is empty.  The second item in the returned submatch
-	// slice is the captured match group.
-	submatches := trailingLeadingDotsDashes.FindStringSubmatch(name)
-	name = submatches[1]
-	return name
+	// If the name is different append a qualifier.
+	return qualifiedName(v1Name, name)
 }
 
 // Convert the v1 name to a standard v3 name with no dots.
-// -  Convert to lowercase
-// -  Convert any char that is not in the set [-a-z0-9] to -
-// -  Remove leading and trailing dashes
 func convertNameNoDots(v1Name string) string {
-	name := strings.ToLower(v1Name)
-	name = nonNameNoDotChar.ReplaceAllString(name, "-")
+	// Normalize the name and then convert dots to dashes.
+	name := names.NormalizeName(v1Name)
+	name = strings.Replace(name, ".", "-", -1)
 
-	// Extract the trailing and leading dashes (there are no dots by this point).
-	// This should always match even if the matched substring is empty.  The second
-	// item in the returned submatc
-	// slice is the captured match group.
-	submatches := trailingLeadingDotsDashes.FindStringSubmatch(name)
-	name = submatches[1]
-	return name
+	// If the name is different append a qualifier.
+	return qualifiedName(v1Name, name)
+}
+
+func qualifiedName(orig, final string) string {
+	// If the name was not modified, just return the unmodified name.
+	if orig == final {
+		return orig
+	}
+	// The name was modified.  Calculate an 8-byte hex qualifier to append.
```

Comment:
Do we have a name character limit? 

> By convention, the names of Kubernetes resources should be up to maximum length of 253 characters

Are we not afraid of going beyond that limit? Suppose we need to handle that case _anyway_ in the migration code... but are we?

Reply:
current plan is to run validate over the converted data.   I hand't planned on attempting to fix up over-long names, but rather reject the config and allow the user the opportunity to fix it up.